### ISSUE TITLE
Problem with global group synchronization

### ISSUE THREAD

**Issue by jcharaoui** [blind-coder/rcmcarddav#136]
I'm encountering a strange problem with the synchronization of a global group. The first synchronisation is always flawless for all Roundcube accounts: all groups are synchronised and each one has the correct amount of members. However, after some time (usually several days) a problem appears whereas at least one group is missing contacts, and an unnamed group appears which contains the missing contacts. In the the `carddav_groups` database, the group entry has an `id`, `abook_id`, `uri` and `cuid`, but strangely the `name`, `vcard` and `etag` columns are empty. The group which has missing contacts is often different across Roundube accounts.

Since the problem only appears after a few days, I'm wondering if the problem is linked to synchronisation processes conflicting with each other. In the logs, I often see `duplicate uri` errors, but haven't been able to determine if/how these are linked to my problem.

My setup is as follows:
Carddav plugin version: latest master
Roundcube: 1.1.4
Carddav server: ownCloud 8.1.4
Number of Roundcube accounts: 12
Synchronisation delay: 10 minutes

**Comment by jcharaoui** [comment]
I've been trying to reproduce this problem manually but I haven't been able to. If you have any ideas, let me know!

**Comment by jcharaoui** [comment]
I should add that there is only one global carddav addressbook configured and that individual carddav addressbooks are disabled.

**Comment by blind-coder** [comment]
Can you post a few of those "duplicate uri" lines?

**Comment by jcharaoui** [comment]
```
[21-Jan-2016 16:22:01 -0500]: <4uel8u50> BACKEND: (dbstore_base) [1062] Duplicate entry '/remote.php/carddav/addressbooks/ltqhm/ltqhm/addressbook-53-6052' for key 'uri'
[21-Jan-2016 16:22:02 -0500]: <4uel8u50> BACKEND: (dbstore_base) [1062] Duplicate entry '/remote.php/carddav/addressbooks/ltqhm/ltqhm/addressbook-1011-60' for key 'uri'
[21-Jan-2016 16:22:03 -0500]: <4uel8u50> BACKEND: (dbstore_base) [1062] Duplicate entry '/remote.php/carddav/addressbooks/ltqhm/ltqhm/addressbook-41-6052' for key 'uri'
```
```
[21-Jan-2016 17:14:52 -0500]: <s53q466b> BACKEND: (get_record_from_carddav) Request for VCF '272300B2-0FFE-4472-883C-1FF90D8B2FEE.vcf' which doesn't exist on the server.
[21-Jan-2016 17:14:54 -0500]: <s53q466b> BACKEND: (get_record_from_carddav) Request for VCF 'DD6CDA6F-58E7-4C84-8CCA-556D760F346F.vcf' which doesn't exist on the server.
[21-Jan-2016 17:14:57 -0500]: <s53q466b> BACKEND: (dbstore_base) [1062] Duplicate entry '/remote.php/carddav/addressbooks/ltqhm/ltqhm/7b2cd3f0-0b3f-43b0-' for key 'uri'
```
The `Request for VCF [...] which doesn't exist on the server.` come up frequently, but I think this is dues to how `use_categories` works, using random etags.

**Comment by blind-coder** [comment]
maybe this is a good time to rewrite the creation of new VCFs to properly use POST...

**Comment by jcharaoui** [comment]
I'm really not familiar with such intricacies but I'd be willing to help test anything that might fix this problem!

**Comment by jcharaoui** [comment]
Oh, and sorry to spam, but I think it's relevant to mention explicitly that I'm using the new `use_categories` parameter with was merged in recently. So this problem might also originate from this patch.

**Comment by jcharaoui** [comment]
I think I've found the origin of my problem, which seems to be caused by the interaction of two bugs.

1. I was running ownCloud 8.1, which is affected by https://github.com/owncloud/contacts/issues/581. Because of this, a number of vCards would get updated (even though they didn't need to), every time this plugin would sync with the server.

2. I noticed that there's nothing preventing two update processes from running concurrently. After activating the debug logs, I noticed that often, multiple `refreshdb_from_server()` processes would run within milliseconds of each other.

I fixed the first issue by upgrading to ownCloud 8.2, and the second with #139

I'm still monitoring the problem and will close the issue once I can confirm it's definately fixed.
